src/users/amoc_users_sup.erl

```diff
@@ -1,60 +1,167 @@
+    Strategy = #{strategy => one_for_one, intensity => 0},
+    {ok, {Strategy, Specs}}.
+
+%% API
+-spec init_storage() -> any().
+init_storage() ->
+    UserSups = lists:reverse(supervisor:which_children(?MODULE)),
+    UserSupPids = [ Pid || {_, Pid, _, _} <- UserSups ],
+    UserSupPidsTuple = erlang:list_to_tuple(UserSupPids),
+    NumOfSupervisors = tuple_size(UserSupPidsTuple),
+    Atomics = atomics:new(1 + NumOfSupervisors, [{signed, false}]),
+    Storage = #storage{user_count = Atomics, sups = UserSupPidsTuple},
+    persistent_term:put(?MODULE, Storage).
+
+-spec clean_storage() -> any().
+clean_storage() ->
+    persistent_term:erase(?MODULE).
+
+-spec count_no_of_users() -> count().
+count_no_of_users() ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:get(Atomics, 1).
+
+-spec incr_no_of_users(non_neg_integer()) -> any().
+incr_no_of_users(SupNum) ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:add(Atomics, 1, 1),
+    atomics:add(Atomics, SupNum + 1, 1).
+
+-spec decr_no_of_users(non_neg_integer()) -> ok.
+decr_no_of_users(SupNum) ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:sub(Atomics, SupNum + 1, 1),
+    case atomics:sub_get(Atomics, 1, 1) of
+        0 ->
+            amoc_controller:zero_users_running();
+        _ ->
+            ok
+    end.
+
+-spec start_child(amoc:scenario(), amoc_scenario:user_id(), any()) -> ok.
+start_child(Scenario, Id, ScenarioState) ->
+    Sup = get_sup_for_user_id(Id),
+    gen_server:cast(Sup, {start_child, Scenario, Id, ScenarioState}).
+
+%% Group all children based on ID to their respective worker supervisor and cast a request with each
+%% group at once. This way we reduce the number of casts to each worker to always one, instead of
+%% depending on the number of users.
+-spec start_children(amoc:scenario(), [amoc_scenario:user_id()], any()) -> ok.
+start_children(Scenario, UserIds, ScenarioState) ->
+    KeyFun = fun(UserId) ->
+                     get_sup_for_user_id(UserId)
+             end,
+    Assignments = maps:groups_from_list(KeyFun, UserIds),
+    CastFun = fun (Sup, Users) ->
+                      gen_server:cast(Sup, {start_children, Scenario, Users, ScenarioState})
+              end,
+    maps:foreach(CastFun, Assignments).
+
+%% Assign a count of children each worker needs to remove
+%% in order to load-balance the request among all workers.
+-spec stop_children(non_neg_integer(), boolean()) -> non_neg_integer().
+stop_children(Count, Force) ->
+    {CountRemove, Assignments} = assign_counts(Count),
+    [ gen_server:cast(Sup, {stop_children, Int, Force}) || {Sup, Int} <- Assignments ],
+    CountRemove.
+
+-spec get_all_children() -> [{pid(), amoc_scenario:user_id()}].
+get_all_children() ->
+    #storage{sups = Sups} = persistent_term:get(?MODULE),
+    All = [ gen_server:call(Sup, get_all_children, infinity) || Sup <- tuple_to_list(Sups) ],
```

Comment:
never expose details of internal `gen_server` protocol, create interface at `amoc_users_worker_sup` module and call it instead.